File: akkademy-db-scala/src/main/scala/com/akkademy/AkkademyDb.scala
package com.akkademy

import akka.actor.Actor
import akka.event.Logging
import com.akkademy.sapi.SetRequest
import scala.collection.mutable.HashMap

class AkkademyDb extends Actor {
  val map = new HashMap[String, Object]
  val log = Logging(context.system, this)

  override def receive = {
    case message: SetRequest => {
      log.info("Received Set request: {}", message)
      map.put(message.key, message.value)
    }
    case o => log.info("received unknown message: {}", o);
  }
}

File: ch2/akkademy-db-java/src/main/java/com/akkademy/AkkademyDb.java
package com.akkademy;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import com.akkademy.messages.GetRequest;
import com.akkademy.messages.KeyNotFoundException;
import com.akkademy.messages.SetRequest;

import java.util.HashMap;
import java.util.Map;

public class AkkademyDb extends AbstractActor {
    protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
    protected final Map<String, Object> map = new HashMap<String, Object>();

    private AkkademyDb(){
        receive(ReceiveBuilder.
                        match(SetRequest.class, message -> {
                            log.info("Received Set request: {}", message);
                            map.put(message.key, message.value);
                            sender().tell(new Status.Success(message.key), self());
                        }).
                        match(GetRequest.class, message -> {
                            log.info("Received Get request: {}", message);
                            Object value = map.get(message.key);
                            Object response = (value != null)
                                    ? value
                                    : new Status.Failure(new KeyNotFoundException(message.key));
                            sender().tell(response, self());
                        }).
                        matchAny(o ->
                                sender().tell(new Status.Failure(new ClassNotFoundException()), self())
                        ).build()
        );
    }
}